Editor: use site locale in editor iframe

ramonjd opened the pull request:
## What? How?

(Only partly) fixes https://github.com/WordPress/gutenberg/issues/52777

A first attempt at settings the lang and dir attributes in the editor iframe.

This PR doesn't look at translations of theme templates.

This PR

- creates user and site locale values in block editor settings
- fetches them in the editor iframe
- sets the `lang` and `dir` attributes of the html element in the iframe
- adds the `rtl` classname to the iframe body tag if necessary
- adds a CSS rule so that block placeholders have the right `dir` (the user locale)

## Why?

So that the editor canvas correctly reflects the current site locale and not the user/editor locale.

## Testing Instructions

## Editor canvas

To test, your user locale should be different to the site locale. By default the user locale is set to the site locale.

Edit your user settings over at `/wp-admin/users.php`, and set a LTR language, such as English or German, as your user locale. 

Set an RTL language, such as Arabic or Hebrew,  as the site locale  in /wp-admin/options-general.php

In the Post and Site editors, check that the `<html />` element contains the correct `lang` and `dir` attributes (from the site).

Now swap the languages: use the RTL language as your user locale and the LTR as your site locale.

Check that the `<html />` element contains the correct `lang` and `dir` attributes (from the site).

## Block/pattern previews

This change should also affect block/pattern previews. To test, either check a theme's pattern library in the Site Editor, insert a Query Block and view the suggested patterns, or open the block or pattern library from the block inserter.

Check that the previews have the correct locale attributes. 

## Screenshots or screencast <!-- if applicable -->

User locale: German (de_DE)
Site locale: Hebrew (he_IL)

<img width="996" alt="Screenshot 2024-07-24 at 1 24 37 PM" src="https://github.com/user-attachments/assets/c35b38c9-e811-4e45-910e-58216e2286d6">

### Patterns

User locale: English (en_US)
Site locale: Arabic (ar)

<img width="1349" alt="Screenshot 2024-07-25 at 11 28 22 AM" src="https://github.com/user-attachments/assets/eeafdde9-e67d-41b8-871d-38aa2b35391a">


## Related

- https://github.com/sabernhardt/gutenberg/pull/2

ramonjd committed:
First commit: testing passing locale to block editor settings.

ramonjd committed:
Lint and force 'ltr'

ramonjd commented on lib/block-editor-settings.php:
```diff
@@ -146,6 +146,47 @@ function gutenberg_get_block_editor_settings( $settings ) {
+	$settings['locale'] = array(
+		'site' => array(
+			'lang'  => $current_site_locale,
+			'isRTL' => $current_site_is_rtl,
```
This is a first pass so everything is up for debate. 

Another option is to just pass the site locale data, e.g.,

```php
	$settings['siteLocale'] = array(
        	'lang'  => $current_site_locale,
        	'isRTL' => $current_site_is_rtl,
	);
```

A few things occurred to me:

- `settings.isRTL` already exists. In the editor, it's the current user locale. Should we ignore it, or add related propertes, e.g., `settings.lang`, `settings.siteLang`, and `settings.siteIsRTL`. Seems a bit clunky.
- Rather than `isRTL: boolean`, what about `dir: 'ltr'|'rtl'`?
- Are block editor settings the right place for this data?

ramonjd replied:
@youknowriad What do you reckon about the locale data model here (and using it in block editor settings)?

Is there a better way to pass site and user locale data to the editor?

I looked at extending the site settings (`/wp/v2/settings`) endpoint, but it appears limited to site option values.

youknowriad replied:
> I looked at extending the site settings (/wp/v2/settings) endpoint, but it appears limited to site option values.

This would have been my preference personally as this is just a site option (whether it's stored as an option or not is not really important IMO). cc @TimothyBJacobs maybe

ramonjd replied:
Yeah, it'd be a neat place to house this stuff.

In that endpoint, we'd add a `is_rtl: true` or `dir: 'rtl'` or whatever, that corresponds to the **site locale**.

The only conceptual agony I was going through was that this value can't be updated in the POST like the others, so it isn't neat, but the world is messy so I can get over it. Keen to hear what Timothy thinks.

Here's the output of the GET request the record:

`await wp.data.resolveSelect( 'core' ).getSite()`

<details>

<summary>Respone</summary>

```json

{
    "title": "gutenberg",
    "description": "",
    "url": "http://localhost:8888",
    "email": "wordpress@example.com",
    "timezone": "",
    "date_format": "F j, Y",
    "time_format": "g:i a",
    "start_of_week": 1,
    "language": "ar",
    "use_smilies": true,
    "default_category": 1,
    "default_post_format": "0",
    "posts_per_page": 10,
    "show_on_front": "posts",
    "page_on_front": 0,
    "page_for_posts": 0,
    "default_ping_status": "open",
    "default_comment_status": "open",
    "site_logo": null,
    "site_icon": 0
}

```

</details>

TimothyBJacobs replied:
This is exactly what the index is really for. The `rest_index` filter can be used.

ramonjd replied:
Thanks for confirming @TimothyBJacobs I'll take a look at using the [filter](https://developer.wordpress.org/reference/hooks/rest_index/) 🙇🏻

ramonjd commented on packages/block-editor/src/components/iframe/index.js:
```diff
@@ -112,14 +112,18 @@ function Iframe( {
+			return {
+				resolvedAssets: settings.__unstableResolvedAssets,
+				isPreviewMode: settings.__unstableIsPreviewMode,
+				siteLocale: settings.locale?.site,
```
Pulling it from the store for convenience, however I'm thinking that the local could be passed to the `Iframe` component as a prop from the consuming components.

This would decouple the Iframe from block settings, and means that the locale can be switched programmatically from above. 🤷🏻

ellatrix replied:
The iframe component is not a public API I think so it probably doesn't matter.

ramonjd commented on packages/block-editor/src/components/iframe/index.js:
```diff
@@ -131,83 +135,88 @@ function Iframe( {
-			for ( const compatStyle of getCompatibilityStyles() ) {
-				if ( contentDocument.getElementById( compatStyle.id ) ) {
-					continue;
+				if ( ! siteLocale ) {
```
Just a placeholder condition, and probably only useful if the locale is passed as a prop.

If no external site locale data is returned from the settings, pull the `dir` from the parent as per usual.

ellatrix replied:
Maybe just remove it here and let's set it in the html element below? Adding dependencies could have some weird effects here, let's avoid that

ellatrix replied:
It's also nice if everything is in one place

ramonjd replied:
Good call, thanks for the feedback!

ellatrix replied:
I don't think this is resolved yet?

ramonjd replied:
Oh I was playing with body classes and forgot I didn't push the commit removing the dependency. Thanks for flagging.

ramonjd reviewed (commented):
This is just a test and doesn't look at anything aside from passing the correct locale attributes to the editor.

There are a few future rabbit holes. 

For example, if the user locale is LTR, should block placeholders be `direction:ltr`?

<img width="759" alt="Screenshot 2024-07-24 at 1 58 10 PM" src="https://github.com/user-attachments/assets/903d1ff1-7fca-49e6-8678-9d4d9e37872d">

Core block assets should be loaded for both RTL and LTR? But isolated to their corresponding window contexts 🙃 

https://github.com/WordPress/gutenberg/blob/af2e21ad7aeac9c834325a23ad77718468c0bf00/lib/blocks.php#L255-L255

I only have a bit of time to devote to this, so happy to have folks nudge the ship along.

sabernhardt commented on packages/block-editor/src/components/iframe/index.js:
```diff
@@ -131,83 +135,88 @@ function Iframe( {
+				// Ideally ALL classes that are added through get_body_class should
+				// be added in the editor too, which we'll somehow have to get from
+				// the server in the future (which will run the PHP filters).
+				setBodyClasses(
```
Could you add the `rtl` body class when the site language direction is RTL?

[My attempt](https://github.com/sabernhardt/gutenberg/pull/2/files#diff-ab17abd38e687edd56573694ae3017e105e96b319571c190e9d4705c1872bc9cR160-R161) for the closely related issue #54150 was simpler but inadequate because it used `ownerDocument` (which has the profile language, not site language).

ramonjd replied:
I'll check it out, thanks for pointing it out.

ramonjd committed:
Adding WP's `rtl` to the body tag
Tentatively adding a rule to ensure that block placeholders, which are part of the editor UI, conform to the directionality of the user locale

ramonjd commented on packages/block-editor/src/components/iframe/index.js:
```diff
@@ -280,6 +287,13 @@ function Iframe( {
+				Block placeholders are "technically" part of the editor UI,
+				and should therefore match the editor UI's directionality.
+			*/
+			.block-editor-block-list__layout .components-placeholder { 
```
This change is optional (follow up?), and there's probably a better way.

The CSS rule here is in lieu of setting the `dir` inside the individual blocks.

With a site lang of `ar` and a user lang of `en`:

 | Before | After |
|--------|--------|
| <img width="396" alt="Screenshot 2024-07-25 at 11 27 36 AM" src="https://github.com/user-attachments/assets/c8a87f04-6d41-4933-9f13-a774433a2041"> | <img width="378" alt="Screenshot 2024-07-25 at 11 27 45 AM" src="https://github.com/user-attachments/assets/61a9feab-2773-4e69-91d6-d09f5519ed5f"> | 

With a site lang of `en` and a user lang of `ar:

 | Before | After |
|--------|--------|
|  <img width="369" alt="Screenshot 2024-07-25 at 11 42 05 AM" src="https://github.com/user-attachments/assets/47f280f5-83a6-453b-a335-938182afa793"> | <img width="380" alt="Screenshot 2024-07-25 at 11 41 36 AM" src="https://github.com/user-attachments/assets/517a58c8-f352-43f7-a744-a305fdc652d2"> |

ramonjd replied:
Maybe we could compose `<Placeholder />` for blocks and do all the locale trickery in the higher order component?

E.g., `const BlockPlaceholder = withLocale( Placeholder )`, or whatever it is

ramonjd committed:
Add e2e tests

youknowriad commented on lib/block-editor-settings.php:
```diff
@@ -146,6 +146,47 @@ function gutenberg_get_block_editor_settings( $settings ) {
+
 	return $settings;
 }
 add_filter( 'block_editor_settings_all', 'gutenberg_get_block_editor_settings', 0 );
```
Why is this hook in `lib/block-editor-settings.php` and not in `wp/6.7` folder...

ramonjd replied:
No reason, mainly due to being distracted by the bigger questions :) I'll move it. 👍🏻

youknowriad commented on packages/block-editor/src/components/iframe/index.js:
```diff
@@ -112,14 +112,19 @@ function Iframe( {
-			isPreviewMode: settings.__unstableIsPreviewMode,
-		};
-	}, [] );
+	const { resolvedAssets, isPreviewMode, siteLocale, userLocale } = useSelect(
```
Why do we need both siteLocale and userLocale? For me userLocale is just `isRTL` so we don't really need a new setting there.

ramonjd replied:
Good question.

The block editor should know about both site locale and user locale because:

1. it needs both the site lang and site RTL value to add the requisite attributes to the document (so that the content matches the frontend)
2. it should have the user locale available in case the editor canvas needs to render any editor-specific UI, such as placeholders.

On point 2, I could be wrong in my assumption, but my instinct tells me that the block placeholders are part of the editor UI, and therefore, the locales should be consistent.

 | Before | After |
|--------|--------|
| <img width="396" alt="Screenshot 2024-07-25 at 11 27 36 AM" src="https://github.com/user-attachments/assets/c8a87f04-6d41-4933-9f13-a774433a2041"> | <img width="378" alt="Screenshot 2024-07-25 at 11 27 45 AM" src="https://github.com/user-attachments/assets/61a9feab-2773-4e69-91d6-d09f5519ed5f"> | 

The iframe may not need to access userLocale here at all, but here I just wanted to provide an example of how to make the placeholder UI conform to the user locale (and not the site). See [above](https://github.com/WordPress/gutenberg/pull/63883#discussion_r1692110347).

I'm thinking a HOC placeholder in block-editor would be a better place to add the `lang` and `dir` attributes to the `<Placeholders />`, not the funky CSS I have in the latest commit of this PR. Anyway, that would be a follow up.

youknowriad replied:
No, my question was simpler than that :P. We already have `wp.i18n.isRTL` function to know whether the user local is rtl or not, we don't need a new setting for that.

ramonjd replied:
You know I love typing :trollface: 

I think I [remember realizing we just needed the site locale.](https://github.com/WordPress/gutenberg/pull/63883#discussion_r1689062491)

So, yeah, good point. "Locale" comprises language too, but `wp.i18n.getLocaleData()[""].lang` should get us the value for the lang attribute too should we need it. 👍🏻

youknowriad reviewed (commented):
Do we have an idea about the classic editor and how it handles this case?

Also the other questions I have (maybe things to address separately):

 - we have block styles `style.css` and `editor.css` and `theme.css` and If I'm not wrong each one of these has "rtl" versions. I'm guessing style and theme need to use the "site locale" but it's unclear what "editor" should be using.
 - Do we have any support for automatic RTL for theme.json generated styles and how does it impact this PR?
 - What about the uniframed block editor, do we ignore that and consider that most people should migrate to iframed editor eventually?

ellatrix commented on packages/block-editor/src/components/iframe/index.js:
```diff
@@ -264,7 +271,7 @@ function Iframe( {
 	// content.
 	const html = `<!doctype html>
-<html>
+<html lang="${ siteLocale?.lang }" dir="${ siteLocale?.isRTL ? 'rtl' : 'ltr' }">
```
Here's another idea: what if the initial HTML as a whole is a setting? This all of this stuff is customisable from outside the iframe component.

ramonjd committed:
Move settings filter to compat/6.7

ramonjd commented:
> Do we have an idea about the classic editor and how it handles this case?

It uses the user locale as far as I can see. There are a couple of related tickets: 

See:

- https://core.trac.wordpress.org/ticket/48996
- https://core.trac.wordpress.org/ticket/40715

> we have block styles style.css and editor.css and theme.css and If I'm not wrong each one of these has "rtl" versions. I'm guessing style and theme need to use the "site locale" but it's unclear what "editor" should be using.
> Do we have any support for automatic RTL for theme.json generated styles and how does it impact this PR?

I was discussing the RTL CSS generation/enqueuing question with @andrewserong today, and yeah, it's something that needs to be address, but in a follow up. 

In my mind whatever is on the editor canvas — the content that represents what we see on the frontend — should follow the site locale.

The editor UI should, in theory, do what it does now and use the current user locale (if set).


This PR deals with HTML semantics pretty much.


> What about the uniframed block editor, do we ignore that and consider that most people should migrate to iframed editor eventually?

I thought about that for 1 second, and concluded it should follow whatever the "app" does, so the user locale as the primary fallback.

All the above sounds easy when I write it down, but I haven't looked too deeply into the how yet. 

I haven't had much time to spend on this, which probably shows, but I think it's important enough to fix so I'll plug away at it when I can.

youknowriad commented:
Yeah, for me the most challenging thing is going to be the CSS loading and how to choose whether we pick RTL or not ... As it's not clear for some third-party CSS files whether they're targetting content or placeholders...

ramonjd committed:
Remove user locale and use current locale

ramonjd committed:
Making things backwards compatible by defaulting to ownerdocument attributes.

ramonjd committed:
Removed placeholder rule. Notes:

Block placeholders are "technically" part of the editor UI, and should therefore match the editor UI's directionality.
! A better approach might be to create a HOC wrapping Placeholder in @wordpress/block-editor that sets the directionality and lang of the Placeholder component

swissspidy commented on lib/compat/wordpress-6.7/block-editor.php:
```diff
@@ -0,0 +1,57 @@
+		}
+	}
+
+	// @TODO confirm if `lang` just needs to be added to the `/wp/v2/settings` response`.
```
`language` is actually already part of that.

ramonjd replied:
Sorry. That was a typo on my part. It should be "confirm if `is_rtl`" pursuant to [this comment](https://github.com/WordPress/gutenberg/pull/63883#discussion_r1692852562)

swissspidy commented on lib/compat/wordpress-6.7/block-editor.php:
```diff
@@ -0,0 +1,57 @@
+
+	if ( $current_user_locale !== $current_site_locale ) {
+		$switched_locale = switch_to_locale( $current_site_locale );
+		if ( is_locale_switched() && $switched_locale ) {
```
You don't need `is_locale_switched()`. Just the return value of `$switched_locale` is enough.


```suggestion
		if ( $switched_locale ) {
```

ramonjd replied:
Oh great, thanks!

ramonjd committed:
Remove is_locale_switched() check and update comment to reflect intented property

ramonjd committed:
Sending language and rtl through from the site index REST response.

Passing locale data down to the iframe from the editor package instead of through the block settings.

ramonjd committed:
Removing block editor settings test for now.

ramonjd committed:
Removed unused arg

ramonjd committed:
LINTY

ramonjd commented:
> Wondering what's needed to bring this closer to finish line.

Probably a way to ensure templates/template parts and other site content is correctly translated into the site language.

I'm not sure the best way to go about it. Maybe sending a `_locale_ value to the APIs.

Also, loading any `-rtl` css into the iframe.

I don't have much time to carry on with this unfortunately, but will keep it on the back burner unless someone decides to pick it up.

ramonjd commented:
Closing for now. I'll come back later if this feature is required and time permits.
